### ISSUE TITLE
The model can not  make prediction.

### ISSUE THREAD

**Issue by bigbigbiubiu** [ultralytics/yolov5#7196]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

 Recently I run train.py file, the result runs-train-exp, train_batch0.jpg can output the marked 0, 1, 2... categories and tabs. val_batch0_labels.jpg can mark out boxes and real category names, but there are neither boxes nor category output in the val_batch0_pred.jpg. Putting the resulting best.pt into the detect.py, only the original picture can be output, but the boxes and categories cannot be detected. What is the reason for this?

**Comment by bigbigbiubiu** [comment]
![image](https://user-images.githubusercontent.com/102585167/160578357-460e8d8c-b864-42e6-8656-8e65edbff339.png)
![image](https://user-images.githubusercontent.com/102585167/160578558-64e04d7f-4d7a-4ff5-8619-daac570fac96.png)

**Comment by glenn-jocher** [comment]
@bigbigbiubiu 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by yuvalmay30** [comment]
Hi
The same happened to me too, how did you solve this issue?

**Comment by glenn-jocher** [comment]
@yuvalmay30 @bigbigbiubiu there is a known bug with Conda/Windows/CUDA11 that causes zero detections. If you run `python detect.py` and see no detections then you may need to downgrade to CUDA 10 to solve the issue, or use one of our premade environments:

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

**Comment by yuvalmay30** [comment]
Thanks for the quick answer @glenn-jocher  !
Notes:  
* I use the Google Colab environment supplied by this repo
* I run `detect.py` and I can see predictions
* I run **`val.py` but can't see predictions or labels**

In order to solve this what do I need to change in the notebook?

**Comment by glenn-jocher** [comment]
@yuvalmay30 detect.py is using FP32 inference, val.py and train.py are using mixed precision (AMP) inference, which seems to clash with CUDA 11 in certain environments, but I've never had any issues with Colab, i.e.:

<img width="995" alt="Screenshot 2022-04-04 at 14 36 00" src="https://user-images.githubusercontent.com/26833433/161544833-087149bd-d603-46e5-809c-925ad859e78d.png">

If your custom model is not producing val results then it's simply been poorly trained and there is no error.

**Comment by yuvalmay30** [comment]
I ran a sanity check:
I trained a model on my custom dataset for like 10 epochs (really poor training).
I then ran `detect.py` with really low `conf` and `iou` and got the following predictions (unsurprisingly):

![d7nb4mv-8c131823-7214-4db7-ab88-14a65496864c_jpg rf 43fb63f8a13a5f6ea5e60295985b5433](https://user-images.githubusercontent.com/83128966/161568871-59fda7af-b248-4240-963c-71da41f03957.jpg)


However, then I ran `val.py` with the same `conf` and `iou` and got 0 predictions:

![val_batch0_pred](https://user-images.githubusercontent.com/83128966/161568753-28189284-f2fc-4628-b531-a966bfe3138f.jpg)

What am I missing?

**Comment by glenn-jocher** [comment]
@yuvalmay30 got it. Then you have a CUDA11/Windows/Conda issue that I mentioned earlier. You can disable AMP in train.py and val.py (and pytorch hub if you use that) to solve the issue, or downgrade your CUDA to CUDA10. See https://github.com/ultralytics/yolov5/issues/7244#issuecomment-1087615221 for another user with the same issue.

**Comment by bigbigbiubiu** [comment]
At the beginning ,I only use 12 pictures ,because I just want to try if I can train my own datasets. Until A scholar told me that the model needs at least 1500 pictures to train parameters ,I increased the number of train images , the problem has been solved .



At 2022-04-04 19:57:21, "yuvalmay30" ***@***.***> wrote:

Hi
The same happened to me too, how did you solve this issue?

**Comment by SeigoMiya** [comment]
Hi
The same happened to me too, I use M1 mac.  how did you solve this issue?

**Comment by bigbigbiubiu** [comment]
Thank you very much. I have solved the problem by increasing the number of pictures on my train and validation set. I think it must be the reason that in the past I used few pictures to train my model .Only through a larger quantities of pictures to train, we can get a better result.



在 2022-03-29 17:42:11，"Glenn Jocher" ***@***.***> 写道：

@bigbigbiubiu👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results.

Most of the time good results can be obtained with no changes to the models or training settings, provided your dataset is sufficiently large and well labelled. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users first train with all default settings before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results we recommend you provide the maximum amount of information possible if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your project/name directory, typically yolov5/runs/train/exp.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

Dataset
Images per class. ≥ 1500 images per class recommended
Instances per class. ≥ 10000 instances (labeled objects) per class recommended
Image variety. Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
Label consistency. All instances of all classes in all images must be labelled. Partial labelling will not work.
Label accuracy. Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
Background images. Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

Model Selection

Larger models like YOLOv5x and YOLOv5x6 will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For mobile deployments we recommend YOLOv5s/m, for cloud deployments we recommend YOLOv5l/x. See our README table for a full comparison of all models.

Start from Pretrained weights. Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the --weights argument. Models download automatically from the latest YOLOv5 release.
python train.py --data custom.yaml --weights yolov5s.pt

                                             yolov5m.pt

                                             yolov5l.pt

                                             yolov5x.pt

                                             custom_pretrained.pt
Start from Scratch. Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty --weights '' argument:
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml

                                                      yolov5m.yaml

                                                      yolov5l.yaml

                                                      yolov5x.yaml
Training Settings

Before modifying anything, first train with default settings to establish a performance baseline. A full list of train.py settings can be found in the train.py argparser.

Epochs. Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
Image size. COCO trains at native resolution of --img 640, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as --img 1280. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same --img as the training was run at, i.e. if you train at --img 1280 you should also test and detect at --img 1280.
Batch size. Use the largest --batch-size that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
Hyperparameters. Default hyperparameters are in hyp.scratch.yaml. We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like hyp['obj'] will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our Hyperparameter Evolution Tutorial.
Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by rinnegann** [comment]
> @yuvalmay30 got it. Then you have a CUDA11/Windows/Conda issue that I mentioned earlier. You can disable AMP in train.py and val.py (and pytorch hub if you use that) to solve the issue, or downgrade your CUDA to CUDA10. See [#7244 (comment)](https://github.com/ultralytics/yolov5/issues/7244#issuecomment-1087615221) for another user with the same issue.

I just set amp to  False in train.py and it worked for me.

**Comment by glenn-jocher** [comment]
@rinnegann glad to hear that worked for you! That's a great solution if you're experiencing issues specifically with mixed precision (AMP) inference. Remember that disabling AMP may result in decreased training speed and increased memory usage.

**Comment by iabahmad** [comment]
> > @yuvalmay30 got it. Then you have a CUDA11/Windows/Conda issue that I mentioned earlier. You can disable AMP in train.py and val.py (and pytorch hub if you use that) to solve the issue, or downgrade your CUDA to CUDA10. See [#7244 (comment)](https://github.com/ultralytics/yolov5/issues/7244#issuecomment-1087615221) for another user with the same issue.
> 
> I just set amp to False in train.py and it worked for me.

how you did this as I am still facing this issue
